pkg/scheduler/admission/podnodeconstraints/admission.go

```diff
@@ -0,0 +1,77 @@
+package podnodeconstraints
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"reflect"
+
+	configlatest "github.com/openshift/origin/pkg/cmd/server/api/latest"
+	"github.com/openshift/origin/pkg/scheduler/admission/podnodeconstraints/api"
+
+	admission "k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+func init() {
+	admission.RegisterPlugin("PodNodeConstraints", func(c kclient.Interface, config io.Reader) (admission.Interface, error) {
+		pluginConfig, err := readConfig(config)
+		if err != nil {
+			return nil, err
+		}
+		return NewPodNodeConstraints(pluginConfig), nil
+	})
+}
+
+func NewPodNodeConstraints(config *api.PodNodeConstraintsConfig) admission.Interface {
+	return &podNodeConstraints{
+		config:  config,
+		Handler: admission.NewHandler(admission.Create),
+	}
+}
+
+type podNodeConstraints struct {
+	*admission.Handler
+	config *api.PodNodeConstraintsConfig
+}
+
+func readConfig(reader io.Reader) (*api.PodNodeConstraintsConfig, error) {
+	if reader == nil || reflect.ValueOf(reader).IsNil() {
+		return &api.PodNodeConstraintsConfig{}, nil
+	}
+
+	configBytes, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+	config := &api.PodNodeConstraintsConfig{}
+	err = configlatest.ReadYAML(configBytes, config)
+	if err != nil {
+		return nil, err
+	}
+	// No validation needed since config is just 2 bools
+	return config, nil
+}
+
+func (oo *podNodeConstraints) Admit(aa admission.Attributes) error {
+	switch {
+	case oo.config == nil,
+		aa.GetResource() != kapi.Resource("pods"),
+		aa.GetKind().Kind != "Pod",
+		aa.GetSubresource() != "",
+		aa.GetOperation() != admission.Create:
+		return nil
+	}
+	pod, ok := aa.GetObject().(*kapi.Pod)
+	if !ok {
+		return admission.NewForbidden(aa, fmt.Errorf("unexpected object: %#v", aa.GetObject()))
+	}
+	if len(pod.Spec.NodeName) > 0 && oo.config.ProhibitNodeName {
```

Comment:
To your point John I think the concern is less that the scheduler is
not in charge, but that the implications of bypassing the scheduler
are vaguely understood.  There are likely scenarios today that are
equally attackable that don't rely on nodeName - the time spent
identifying and closing those gaps (or convincing ourselves those are
safe) is more valuable than the time to close one specific possible
branch.

Kube as a system is designed for nodeName being a targetable field.
Adding this is creating technical debt for future work on custom
scheduling, reschedulers, and carried maintenance.  An assessment of
how the nodes and scheduler identify and cope with abuse is going to
be far more valuable, even before the complexities of this solution
are complete.